File: classify_proposals.py
import os
import re
import argparse
import google.generativeai as genai
from pypdf import PdfReader
from docx import Document

class Palette:
    GREEN = '\033[92m'
    CYAN = '\033[96m'
    RED = '\033[91m'
    RESET = '\033[0m'
    BOLD = '\033[1m'

    @staticmethod
    def colorize(text, color):
        return f"{color}{text}{Palette.RESET}"

def extract_text(filepath):
    text = ""
    try:
        if filepath.lower().endswith('.pdf'):
            reader = PdfReader(filepath)
            # Use first 5 pages for classification to save time/memory,
            # usually sufficient for header/structure analysis.
            for page in reader.pages[:5]:
                extracted = page.extract_text()
                if extracted:
                    text += extracted + "\n"
        elif filepath.lower().endswith('.docx'):
            doc = Document(filepath)
            # First 50 paragraphs
            for para in doc.paragraphs[:50]:
                text += para.text + "\n"
        elif filepath.lower().endswith('.doc'):
             # Cannot easily read .doc without antiword or catdoc.
             # Assuming 'form' in name -> Respond, else Flex?
             # For this script we return empty and let filename heuristics handle it.
             pass
    except Exception as e:
        # print(f"Error reading {filepath}: {e}")
        pass
    return text

_gemini_client_configured = False
_gemini_model = None

def classify_with_gemini(text):
    global _gemini_client_configured, _gemini_model
    api_key = os.environ.get("GEMINI_API_KEY")
    if not api_key:
        return None

    try:
        if not _gemini_client_configured:
            genai.configure(api_key=api_key)
            _gemini_model = genai.GenerativeModel("gemini-1.5-flash")
            _gemini_client_configured = True

        prompt = (
            "You are an expert grant proposal classifier. "
            "Classify the following document text as either 'Flex (Type 2)' or 'Respond (Type 1)'. "
            "If the document contains 'Request for Proposal', 'RFP', 'Questions', 'Guidelines', or 'Narrative', it is likely 'Flex (Type 2)'. "
            "If the document is a 'Form', 'Template', 'Checklist', or 'Application Form' to be filled out, it is likely 'Respond (Type 1)'. "
            "Return ONLY the classification string: 'Flex (Type 2)' or 'Respond (Type 1)'."
            "\n\nDocument Text:\n" + text[:10000] # Limit context window just in case
        )

        response = _gemini_model.generate_content(prompt)
        return response.text.strip()
    except Exception as e:
        # print(f"Gemini API Error: {e}")
        return None

def classify_file(filepath):
    text = extract_text(filepath)

    # Try Gemini first if API key is present
    gemini_result = classify_with_gemini(text)
    if gemini_result and ("Flex" in gemini_result or "Respond" in gemini_result):
        return gemini_result

    # Fallback to Heuristics
    filename = os.path.basename(filepath).lower()
    text = text.lower()

    # --- Heuristic Rules ---

    score_respond = 0
    score_flex = 0

    # 1. Filename Analysis
    if "template" in filename: score_respond += 3
    if "form" in filename:
         # "Application form" in Respond/Application form for GPE grants.docx -> Should be Respond
         score_respond += 2
    if "checklist" in filename: score_respond += 2
    if "chapter" in filename: score_respond += 2 # Based on Respond/Chapter_3.pdf
    if "appendix" in filename: score_respond += 2 # Based on Respond/Appendix...

    if "rfp" in filename: score_flex += 3
    if "question" in filename: score_flex += 3
    if "guidelines" in filename: score_flex += 2
    if "narrative" in filename: score_flex += 2

    # 2. Content Analysis

    # FLEX Indicators
    if "character limit" in text: score_flex += 3 # Found in 'QuestionList type 2.pdf'
    if "request for proposal" in text: score_flex += 2
    if "project overview" in text and "project name*" in text: score_flex += 3 # Specific form style in Flex
    if "collaborate feature" in text: score_flex += 2 # Found in QuestionList (1).pdf
    if "characters maximum" in text: score_flex += 3 # Found in Grant-form-2025-V2C1.pdf

    # RESPOND Indicators
    if "application form" in text:
        # Check context. If it's "Grant Application Form" it might be Respond.
        # But some Flex docs also mention "application form".
        # Let's check for "fillable" or "template" alongside it.
        # If 'application form' is in the text but NOT 'character limit' or 'project overview', maybe Respond?
        score_respond += 1

    if "please note that a session will time out" in text: score_respond += 2 # Found in Respond template
    if "application template" in text: score_respond += 3
    if "chapter 3:" in text or "chapter 2:" in text: score_respond += 2 # Structure of Respond files seems segmented

    # Specific Correction for Guidelines misclassification
    if "grant guidelines" in text and "guidelines" not in filename:
        # Sometimes guidelines are in Respond if they are part of a template pack?
        # But usually 'Guidelines' -> Flex.
        # Let's see: 'Flex/2025-Grant-Application-Rancheria-Fund-Grant-Cycle.pdf' has "GRANT GUIDELINES" in text.
        # It was classified as Respond. Why?
        # Filename '2025-Grant-Application...' -> contains 'Application' (no specific rule), 'Grant'.
        # Text has 'Grant Cycle Guidelines'.
        # Let's boost Flex if "Guidelines" in text significantly.
        score_flex += 1

    # Length/Structure Heuristics (Weak)
    # If text has many "_____" (underscores for filling), it might be a form (Respond)
    if text.count("___") > 10: score_respond += 1

    # Conflict Resolution
    if score_flex > score_respond:
        return "Flex (Type 2)"
    elif score_respond > score_flex:
        return "Respond (Type 1)"
    else:
        # Fallback based on folder contents observation
        # Respond seems to have more "standard" looking docs, Flex has more "QuestionList"
        if "question" in filename or "rfp" in filename:
            return "Flex (Type 2)"

        # If text length is very long and structure is complex (many bullets?), maybe Flex?
        # If it's short and structured, Respond.

        return "Respond (Type 1)" # Default to Respond if unsure

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Classify Grant Proposals')
    parser.add_argument('files', metavar='F', type=str, nargs='+', help='Files to classify')
    args = parser.parse_args()

    flex_count = 0
    respond_count = 0

    print(f"\n{Palette.BOLD}Classifying Proposals...{Palette.RESET}\n")

    for f in args.files:
        result = classify_file(f)

        color = Palette.GREEN if "Flex" in result else Palette.CYAN
        formatted_result = Palette.colorize(result, color)

        if "Flex" in result:
            flex_count += 1
        elif "Respond" in result:
            respond_count += 1

        if os.path.exists(f):
            print(f"📄 {f}: {formatted_result}")
        else:
            # Handle the missing files mentioned by user
            # We still count them as they were classified by name
            error_msg = Palette.colorize("(File not found)", Palette.RED)
            print(f"❌ {f} {error_msg}: {formatted_result}")
    print(f"\n{Palette.BOLD}Summary:{Palette.RESET}")
    print(f"  Flex:    {Palette.colorize(str(flex_count), Palette.GREEN)}")
    print(f"  Respond: {Palette.colorize(str(respond_count), Palette.CYAN)}")
    print("")

File: src/classify_proposals.ts
/**
 * Grant Proposal Classifier - TypeScript/Deno implementation
 *
 * Classifies grant proposal documents as either "Flex (Type 2)" or "Respond (Type 1)"
 * using heuristic rules and optional Gemini AI classification.
 */

import type { GeminiApiResponse, ProposalType } from "./types.ts";

/** Type guard that validates the shape of a Gemini API response. */
function isGeminiApiResponse(value: unknown): value is GeminiApiResponse {
  if (value === null || typeof value !== "object") return false;
  const obj = value as Record<string, unknown>;
  if (!Array.isArray(obj["candidates"])) return false;
  for (const candidate of obj["candidates"] as unknown[]) {
    if (candidate === null || typeof candidate !== "object") return false;
    const c = candidate as Record<string, unknown>;
    if (c["content"] === null || typeof c["content"] !== "object") return false;
    const content = c["content"] as Record<string, unknown>;
    if (!Array.isArray(content["parts"])) return false;
    for (const part of content["parts"] as unknown[]) {
      if (part === null || typeof part !== "object") return false;
      if (typeof (part as Record<string, unknown>)["text"] !== "string") {
        return false;
      }
    }
  }
  return true;
}

/**
 * Classify text using the Gemini AI API.
 * Returns null if the API key is not set or if an error occurs.
 */
export async function classifyWithGemini(
  text: string,
): Promise<ProposalType | null> {
  const apiKey = Deno.env.get("GEMINI_API_KEY");
  if (!apiKey) {
    return null;
  }

  const prompt =
    "You are an expert grant proposal classifier. " +
    "Classify the following document text as either 'Flex (Type 2)' or 'Respond (Type 1)'. " +
    "If the document contains 'Request for Proposal', 'RFP', 'Questions', 'Guidelines', or 'Narrative', it is likely 'Flex (Type 2)'. " +
    "If the document is a 'Form', 'Template', 'Checklist', or 'Application Form' to be filled out, it is likely 'Respond (Type 1)'. " +
    "Return ONLY the classification string: 'Flex (Type 2)' or 'Respond (Type 1)'." +
    "\n\nDocument Text:\n" +
    text.slice(0, 10000);

  try {
    const url =
      `https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key=${apiKey}`;
    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        contents: [{ parts: [{ text: prompt }] }],
      }),
    });

    if (!response.ok) {
      return null;
    }

    const data: unknown = await response.json();
    if (!isGeminiApiResponse(data) || data.candidates.length === 0) {
      return null;
    }

    const parts = data.candidates[0].content.parts;
    if (parts.length === 0) {
      return null;
    }

    const result = parts[0].text.trim();
    if (result === "Flex (Type 2)" || result === "Respond (Type 1)") {
      return result;
    }
    return null;
  } catch (error: unknown) {
    if (error instanceof Error) {
      console.error("Gemini API error:", error.message);
    }
    return null;
  }
}

/**
 * Classify a proposal document by its filename and text content using heuristic rules.
 */
export function classifyByHeuristics(
  filename: string,
  text: string,
): ProposalType {
  const nameLower = filename.toLowerCase();
  const textLower = text.toLowerCase();

  let scoreRespond = 0;
  let scoreFlex = 0;

  // 1. Filename Analysis
  if (nameLower.includes("template")) scoreRespond += 3;
  if (nameLower.includes("form")) scoreRespond += 2;
  if (nameLower.includes("checklist")) scoreRespond += 2;
  if (nameLower.includes("chapter")) scoreRespond += 2;
  if (nameLower.includes("appendix")) scoreRespond += 2;

  if (nameLower.includes("rfp")) scoreFlex += 3;
  if (nameLower.includes("question")) scoreFlex += 3;
  if (nameLower.includes("guidelines")) scoreFlex += 2;
  if (nameLower.includes("narrative")) scoreFlex += 2;

  // 2. Content Analysis

  // FLEX Indicators
  if (textLower.includes("character limit")) scoreFlex += 3;
  if (textLower.includes("request for proposal")) scoreFlex += 2;
  if (
    textLower.includes("project overview") &&
    textLower.includes("project name*")
  ) {
    scoreFlex += 3;
  }
  if (textLower.includes("collaborate feature")) scoreFlex += 2;

  // RESPOND Indicators
  if (textLower.includes("application form")) scoreRespond += 1;
  if (textLower.includes("please note that a session will time out")) {
    scoreRespond += 2;
  }
  if (textLower.includes("application template")) scoreRespond += 3;
  if (textLower.includes("chapter 3:") || textLower.includes("chapter 2:")) {
    scoreRespond += 2;
  }

  if (
    textLower.includes("grant guidelines") &&
    !nameLower.includes("guidelines")
  ) {
    scoreFlex += 1;
  }

  // Length/Structure Heuristics
  const underscoreMatches = textLower.match(/___/g);
  if (underscoreMatches && underscoreMatches.length > 10) scoreRespond += 1;

  // Conflict Resolution
  if (scoreFlex > scoreRespond) {
    return "Flex (Type 2)";
  } else if (scoreRespond > scoreFlex) {
    return "Respond (Type 1)";
  } else {
    if (
      nameLower.includes("question") ||
      nameLower.includes("rfp")
    ) {
      return "Flex (Type 2)";
    }
    return "Respond (Type 1)";
  }
}

/**
 * Classify a proposal given its filename and text content.
 * Tries Gemini AI first, then falls back to heuristic rules.
 */
export async function classifyProposal(
  filename: string,
  text: string,
): Promise<ProposalType> {
  const geminiResult = await classifyWithGemini(text);
  if (geminiResult !== null) {
    return geminiResult;
  }

  return classifyByHeuristics(filename, text);
}
